### ISSUE TITLE
bbapi_wdt: select PwrCtrl IO failed with: 0x708

### ISSUE THREAD

**Issue by bebac** [Beckhoff/BBAPI#6]
bbapi_wdt fails with kernel 4.18 gcc 8.2

[    0.000000] Linux version 4.18.16-rt9+ (vagrant@mainctrl-vagrant) (gcc version 8.2.0 (GCC)) #1 SMP PREEMPT RT Thu Mar 14 06:39:23 UTC 2019
.
[    0.000000] DMI: Beckhoff Automation GmbH & Co. KG CBxx55/56/CB3056, BIOS Bxx56_AMT - 1.61 10/24/2016
.
[    6.961195] bbapi: Beckhoff BIOS API Driver, 1.5
[    7.225789] bbapi: BIOS found and copied from: (____ptrval____) + 0xb9600 | 0
[    7.232911] bbapi: platform has no display or is not supported
[    7.240983] bbapi_wdt: Beckhoff BIOS API watchdog driver, 0.5 (nowayout=1)
[    8.259127] bbapi_wdt: select PwrCtrl IO failed with: 0x708
[    8.270786] bbapi_wdt: select PwrCtrl IO failed with: 0x708

Here it is on our current working system:
[    0.000000] Linux version 4.1.42-rt50+ (vagrant@mainctrl-vagrant) (gcc version 6.2.0 (GCC) ) #1 SMP PREEMPT RT Thu Jan 3 09:08:48 UTC 2019
.
[    0.000000] DMI: Beckhoff Automation GmbH & Co. KG CBxx55/56/CB3056, BIOS Bxx56_AMT - 1.61 10/24/2016
.
[    7.023221] bbapi: Beckhoff BIOS API Driver, 1.5
[    7.258125] bbapi: BIOS found and copied from: ffffc90002200000 + 0xb9600 | 0
[    7.265241] bbapi: platform has no display or is not supported
[    7.272543] bbapi_wdt: Beckhoff BIOS API watchdog driver, 0.5 (nowayout=1)

**Comment by pbruenn** [comment]
Do you have the kernel config files at hand? At least a diff between old and new config would be quit helpful.

**Comment by bebac** [comment]
Yes, of course. I should also mention that it is not the latest version of the code. I have attached the code too
[kernel_config_old.txt](https://github.com/Beckhoff/BBAPI/files/2969849/kernel_config_old.txt)
[kernel_config_new.txt](https://github.com/Beckhoff/BBAPI/files/2969852/kernel_config_new.txt)
[bbapi.tar.gz](https://github.com/Beckhoff/BBAPI/files/2969854/bbapi.tar.gz)

**Comment by pbruenn** [comment]
Thanks, I didn't had time to setup a test system yet, but from looking into the configs my first guess would be:
CONFIG_I2C_SMBUS=y
CONFIG_I2C_I801=y
That's interesting because the PwrCtrl is connected by SMBUS and 0x708 means bus busy. So my guess is, there is some Linux driver using the bus.
Is the error constantly reproduce able on 4.18 and never occurs on 4.1?
Did you tried the latest bbapi driver version?

**Comment by bebac** [comment]
Interesting, not sure why CONFIG_I2C_SMBUS got enabled in the new config. I will look into that. It is constantly reproduce-able at least the handful of times i tried to reboot. 

I did try the latest version, but that got stuck on boot. I suspect it might be related somehow

**Comment by bebac** [comment]
Not sure why I closed the issue

**Comment by bebac** [comment]
Tried with:
```
@@ -1849,7 +1849,7 @@ CONFIG_I2C_COMPAT=y
 # CONFIG_I2C_CHARDEV is not set
 # CONFIG_I2C_MUX is not set
 CONFIG_I2C_HELPER_AUTO=y
-CONFIG_I2C_SMBUS=y
+# CONFIG_I2C_SMBUS=y
 CONFIG_I2C_ALGOBIT=m

```
Makes no difference

**Comment by bebac** [comment]
I have tried to trace smbus and i2c events without any luck though. Either i am doing something wrong or it never gets to the trace points.

    modprobe bbapi
    modprobe bbapi_wdt

    mount -t debugfs none /sys/kernel/debug/
    echo 1 > events/smbus/enable
    echo 1 > events/i2c/enable

Then try to kick the watchdog

    echo 1 > /dev/watchdog

This will result in the log

   [  204.101380] bbapi_wdt: select PwrCtrl IO failed with: 0x708

But no smbus or i2c events are logged

**Comment by bebac** [comment]
I have tried the latest version, but target freezes when loading bbapi_wdt

**Comment by pbruenn** [comment]
Today, I tried latest bbapi with Linux 5.0 on CX2030 and CB3056 without issues.
CONFIG_SMSC_SCH311X_WDT=m sound promising, but if you had that driver loaded the error message would look different ("a legacy watchdog module is probably present").
So I am still trying to reproduce your issue. Currently building 4.18.16-rt9...

**Comment by pbruenn** [comment]
Your kernel_config_new.txt didn't boot on my CX2030. My config:
[kernel_config_new-CX2030.txt](https://github.com/Beckhoff/BBAPI/files/2986750/kernel_config_new-CX2030.txt)
had no issues with 4.18.16-rt9. I will try to merge mine into your direction.
Well I had issues with i915 driver and rtmutex, but bbapi_watchdog was working fine.
i915 will be the first thing I disable in the next round

**Comment by bebac** [comment]
I got the modified version (the bbapi.tar.gz from earlier) working. I think it is something with the I2C_SUPPORT. Previously I manually disabled just CONFIG_I2C_SMBUS, but now I used the menuconfig and set I2C_SUPPORT to be a module as it apparently cannot be disabled completely from menuconfig. That set a bunch of other things to be modules, I am not sure of the impact of that.

I will try out the latest version with I2C_SUPPORT set to m

BTW this is on a CB3056. That is also the one where I first saw the problem. However some of the experiments from earlier was done on a CB3064 that I was testing out.

**Comment by bebac** [comment]
Did some more testing with I2C_SUPPORT=m

Tested CB3056 with old modified bbapi (api=1.5, wdt=0.5) and new modified bbapi (api=2.0, wdt=0.5) and they both work

Did the same test with CB3064. Here, strangely enough, only the old modified bbapi works. With the new version target freezes as soon as the watchdog daemon is launched. The modules actually loads fine. It is not until the watchdog daemon start kicking the watchdog that target hangs.

**Comment by pbruenn** [comment]
Okay, then I try to get some CB3064 hardware.
Could you please open a new issue like "bbapi_wdt: freezes on CB3064, when watchdog fires" just to make it easier for others to follow up (most recent kernel config is a bonus).

Having the full config here would be great, too. Making it easier for other readers to follow the "I2C_SUPPORT=m" trick.
Many thanks in advance,
Patrick

**Comment by bebac** [comment]
Here is the full working 4.18.16-rt9 kernel config
[kernel_config_bbapi_wdt_working.txt](https://github.com/Beckhoff/BBAPI/files/2989200/kernel_config_bbapi_wdt_working.txt)

I will double check this configuration with the new CB3064 and open a separate issue for that.

Thank you very much for the support

**Comment by pbruenn** [comment]
I think the only remaining issue was #7 which is now fixed on master.
If you disagree, feel free to reopen or add a new issue.
